<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Работа-с-пропусками-и-аномалиями-в-данных" data-toc-modified-id="Работа-с-пропусками-и-аномалиями-в-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Работа с пропусками и аномалиями в данных</a></span></li><li><span><a href="#Замена-вещественного-типа-данных-на-целочисленный" data-toc-modified-id="Замена-вещественного-типа-данных-на-целочисленный-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена вещественного типа данных на целочисленный</a></span></li><li><span><a href="#Дубликаты" data-toc-modified-id="Дубликаты-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Дубликаты</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Определение-зависимостей" data-toc-modified-id="Определение-зависимостей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Определение зависимостей</a></span><ul class="toc-item"><li><span><a href="#Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Зависимость между количеством детей и возвратом кредита в срок</a></span></li><li><span><a href="#Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Зависимость между семейным положением и возвратом кредита в срок</a></span></li><li><span><a href="#Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Зависимость между уровнем дохода и возвратом кредита в срок</a></span></li><li><span><a href="#Зависимость-между-целью-получения-кредита-и-его-возвратом-в-срок" data-toc-modified-id="Зависимость-между-целью-получения-кредита-и-его-возвратом-в-срок-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Зависимость между целью получения кредита и его возвратом в срок</a></span></li></ul></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Итоги исследования</a></span></li></ul></div>

# Исследование надёжности заёмщиков

**Цель исследования** - определить, существуют ли следующие зависимости:
1. Зависимость между количеством детей и возвратом кредита в срок.
2. Зависимость между семейным положением и возвратом кредита в срок.
3. Зависимость между уровнем дохода и возвратом кредита в срок.
4. Зависимость между целью получения кредита и его возвратом в срок.

**Ход исследования**

Входные данные от банка — статистика о платёжеспособности клиентов. О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных.

Необходимо проверить данные на ошибки и оценить их влияние на исследование. Затем, на этапе предобработки нужно исправить самые критичные ошибки данных.

Таким образом, исследование пройдёт в три этапа:

1. Обзор данных.
2. Предобработка данных.
3. Определение зависимостей.

## Обзор данных

Импортируем библиотеку `pandas`.

In [1]:
import pandas as pd

Читаем файл `data.csv` и сохраняем его в переменной `data`:

In [2]:
try:
    data = pd.read_csv(r'C:\Users\м\Documents\Data\data.csv')
except:
    data = pd.read_csv('/datasets/data.csv')

Выводим первые десять строк таблицы `data` на экран:

In [3]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получаем общую информацию о таблице:

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов. Согласно документации к данным:
- `children` — количество детей в семье
- `days_employed` — общий трудовой стаж в днях
- `dob_years` — возраст клиента в годах
- `education` — уровень образования клиента
- `education_id` — идентификатор уровня образования
- `family_status` — семейное положение
- `family_status_id` — идентификатор семейного положения
- `gender` — пол клиента
- `income_type` — тип занятости
- `debt` — имел ли задолженность по возврату кредитов
- `total_income` — ежемесячный доход
- `purpose` — цель получения кредита

Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.

Нужно устранить проблемы в данных.

## Предобработка данных

### Работа с пропусками и аномалиями в данных
Посчитаем, сколько в таблице пропущенных значений:

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В обоих случаях пропущено 2174 значения. Для датафрейма `data`, в котором 21525 строк, это составляет ~10,1% от всех значений. Это довольно большая доля, поэтому мы не будем удалять строки с пропущенными значениями, это может сказаться на результатах исследования.

Из таблицы видно, что в столбце `days_employed` есть отрицательные значения. Похоже, что при внесении данных имел место человеческий фактор.  Поскольку трудовой стаж не может быть отрицательным, перезапишем значения в положительные с помощью встроенной функции `abs()`: 

In [6]:
data['days_employed'] = data['days_employed'].abs()

Так как данные в столбцах `days_employed` и `total_income` представлены количественными переменными, пропуски в них нужно заполнить характерными значениями. Даже из первых десяти строк таблицы видно, что в столбце `days_employed` есть очень большие значения, превышающие даже возраст человека. Возможно, по ошибке данные были записаны не в днях, а в часах. Чтобы избежать влияния на результаты исследования сильно выделяющихся значений, заменим пропуски медианным значением. 

Создадим цикл, который пройдет по всем значениям столбцов `days_employed` и `total_income`, рассчитает медианное значение по каждому из столбцов и заполнит ими пропущенные значения в этих столбцах: 

In [7]:
columns_to_replace = ['days_employed', 'total_income'] #создаем список из названий двух столбцов
for column in columns_to_replace:                      #перебор значений в цикле, расчет медианного значения и заполнение 
    avg = data[column].median()                        #им пропусков  
    data[column] = data[column].fillna(avg) 

Убедимся, что в таблице не осталось пропущенных значений:

In [8]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Посмотрим, есть ли в данных другие аномалии. Посмотрим на список уникальных значений в столбце `children`:

In [9]:
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Количество детей не может быть отрицательным - очевидно, что это ошибка при вводе данных. Вероятнее всего, значение 20 детей - это тоже человеческая ошибка. Устраним эти проблемы. 

In [10]:
data['children'] = data['children'].abs()    #приведем все значения столбца к положительным

data = data.loc[data['children'] != 20]      #так как число строк, где кол-во детей равно 20, составляет несущественную долю  
                                             #от общего числа строк, удалим их из датафрейма   

Взглянем также на уникальные значения столбца `dob_years`:

In [11]:
data['dob_years'].value_counts()

35    615
40    605
41    605
34    600
38    597
42    594
33    579
39    572
31    558
36    553
44    545
29    543
48    537
30    537
37    533
50    511
43    511
32    508
49    505
28    503
45    494
27    491
52    483
56    482
47    480
54    478
46    472
58    461
57    459
53    458
51    447
59    442
55    442
26    407
60    376
25    356
61    354
62    351
63    269
64    264
24    263
23    253
65    194
66    183
22    183
67    167
21    110
0     100
68     99
69     84
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Вывод показал нам, что есть 100 строк с нулевым значением возраста клиента. Так как в данном столбце нет аномальных значений, заполним эти строки средним возрастом клиентов.

In [12]:
#после операции замены возвращаем тип данных int
data['dob_years'] = data['dob_years'].replace(0, data['dob_years'].mean()).astype('int') 

### Замена вещественного типа данных на целочисленный

Для более удобной работы со значениями столбца `total_income` заменим вещественный тип данных `float` на целочисленный `int`:

In [13]:
data['total_income'] = data['total_income'].astype('int') 

### Дубликаты

Посчитаем количество строк-дубликатов в таблице с помощью сочетания методов `duplicated()` и `sum()`, а затем удалим повторяющиеся строки.

In [14]:
data.duplicated().sum()    #считаем кол-во строк-дубликатов

54

In [15]:
#создаем новый датафрейм без дубликатов и обновляем индексацию
data = data.drop_duplicates().reset_index(drop=True) 

Теперь необходимо проверить данные на неявные дубликаты. Уже из первых строк таблицы, которую мы выводили на экран, видно что значения столбца `education` отличаются регистром. Посмотрим на все уникальные значения этого столбца:

In [16]:
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Действительно, в столбце есть дублирующиеся значения, отличающиеся только написанием. Переведем их в нижний регистр и заново посчитаем количество явных дубликатов в датафрейме.  

In [17]:
#перезаписываем значения столбца в нижний регистр и считаем кол-во строк-дубликатов
data['education'] = data['education'].str.lower()         
data.duplicated().sum()                                   

17

Нашлось еще семнадцать строк-дубликатов. Удалим их аналогичным образом - с использованием метода `drop_duplicates()`:

In [18]:
#создаем новый датафрейм без дубликатов и обновляем индексацию
data = data.drop_duplicates().reset_index(drop=True) 

Далее поочередно проверим значения всех столбцов, содержащих данные типа `object` на предмет наличия значений-дубликатов, записанных в разном регистре.

In [19]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [20]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

Мы обнаружили аномальное значение `XNA` в столбце `gender`. Посмотрим, много ли таких значений:  

In [21]:
data.loc[data['gender']=='XNA']['gender'].count()

1

Поскольку значение `XNA` встречается всего один раз, можно предположить, что при записи произошла техническая ошибка. Наличие одной единственной строки с таким значением никак не скажется на результатах исследования, не будем ее удалять. 

In [22]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [23]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Во всех проверенных столбцах не было обнаружено проблем с использованием разных регистров. Однако в последнем столбце `purpose` очень много похожих по смыслу, но по-разному записанных значений. К решению этой проблемы мы вернемся позже.  

**Выводы**

Нами была обнаружена и удалена 71 строка-дубликат. Возможно, дубликаты возникли из-за повторного обращения некоторых клиентов в банк и соответственно повторного внесения их в таблицу. Также дубликаты появились в связи с неучетом регистра при заполнении таблицы.     

Для поиска явных дубликатов нами был использован метод `duplicated()`. Он находит строки, значения которых совпадают во всех столбцах. Сочетая его с методом `sum()`, мы нашли количество строк-дубликатов. Иначе мы бы получили Series со значением True для строк-дубликатов.  
Неявные дубликаты мы нашли с помощью метода `unique()`, он возвращает список уникальных значений в столбце. Для того, чтобы найти дубликаты с учетом регистра и привести записи к единообразию, мы вызвали метод `str.lower()`. 
Все дубликаты мы удалили с применением метода `drop_duplicates()`. Он создал очищенный от повторяющихся строк датафрейм. 

### Категоризация данных

Для оптимизации данных разделим таблицу `data` и создадим по ней два «словаря»:

In [24]:
#создаем новый датафрейм со столбцами с уровнем образования клиентов
#и с идентификаторами уровня образования 
data_education = data[['education', 'education_id']]    
data_education.head()                                 

education  education_id
0    высшее             0
1   среднее             1
2   среднее             1
3   среднее             1
4   среднее             1

In [25]:
#создаем новый датафрейм со столбцами с семейным положением
#клиентов и с идентификаторами семейного положения
data_family = data[['family_status', 'family_status_id']]  
data_family.head()                                       

family_status  family_status_id
0   женат / замужем                 0
1   женат / замужем                 0
2   женат / замужем                 0
3   женат / замужем                 0
4  гражданский брак                 1

В созданных таблицах много дубликатов, удалим их, применив знакомую цепочку методов: `drop_duplicates()` и `reset_index()`:

In [26]:
data_education = data_education.drop_duplicates().reset_index(drop=True)
data_education.head(10)                                         #выведем результат на экран

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [27]:
data_family = data_family.drop_duplicates().reset_index(drop=True)
data_family.head(10)                                         #выведем результат на экран

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

Для той же оптимизации данных удалим из исходного датафрейма столбцы `education` и `family_status`, оставим только столбцы с их идентификаторами: 

In [28]:
data = data.drop(['education', 'family_status'], axis = 1)      #создаем новую таблицу без двух указанных столбцов
data.head()                                                     #проверяем внесенные изменения

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

Таким образом мы уменьшили размер исходной таблицы, ускорили процесс фильтрации данных и упростили визуальную работу с датафреймом. В дальнейшем, чтобы проанализировать данные по уровню образования или семейному положению клиентов, не нужно будет запоминать полное словесное выражение, достаточно будет обратиться к идентификатору и по нему получить значение из «словаря». 

Теперь разделим клиентов на категории в зависимости от размеров ежемесячного дохода. Для этого создадим функцию.
Категории распределим следующим образом:
- 0–30000 — E;
- 30001–50000 — D;
- 50001–200000 — C;
- 200001–1000000 — B;
- 1000001 и выше — A.

In [29]:
def income_group(income):
    if income <= 30000:
        return 'E'
    if income <= 50000:
        return 'D'
    if income <= 200000:
        return 'C'
    if income <= 1000000:
        return 'B'
    return 'A'

Создадим в датафрейме отдельный столбец с категориями дохода:

In [30]:
#применяем созданную ранее функцию к значениям столбца 'total_income'
#выводим часть датафрейма на экран

data['total_income_category'] = data['total_income'].apply(income_group)                                                                          
data.head(10)                                                            

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

Создадим также отдельный столбец с категориями покупателей по цели получения кредита. Ранее мы выводили уникальные значения столбца `purpose` на экран и увидели, что в нем много похожих по смыслу значений. Определим следующие категории:
- операции с автомобилем;
- операции с недвижимостью;
- проведение свадьбы;
- получение образования.

Вначале пропишем функцию, которая будет искать ключевые слова в столбце с целью и возвращать название категории:

In [31]:
def purpose_group(purpose):
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    if 'недвижимост' in purpose:
        return 'операции с недвижимостью'
    if 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образован' in purpose:
        return 'получение образования'

А сейчас создадим в датафрейме отдельный столбец с категориями целей получения кредита:

In [32]:
#применяем созданную ранее функцию к значениям столбца 'purpose'
#выводим часть датафрейма на экран

data['purpose_category'] = data['purpose'].apply(purpose_group)                                                                    
data.head(10)                                                        

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы  
5                     B  операции с недвижимостью  
6                     B  операции с недвижимостью  
7                     C     получение образования  
8                     C        проведение свадьбы  
9                     C  операции с недвижимостью

Проверим, по всем ли строкам удалось определить категорию цели получения кредита:

In [33]:
#ищем пропущенные значения в столбце 'purpose_category'
data['purpose_category'].isna().sum()     

0

Теперь, когда все данные очищены от пропусков, дубликатов и в целом приведены в порядок, можно приступать к поиску взаимосвязей и ответу на вопросы.

## Определение зависимостей

### Зависимость между количеством детей и возвратом кредита в срок

Для того, чтобы проверить существует ли зависимость между количеством детей у клиентов банка и возвратом кредита в срок, построим сводную таблицу:

In [34]:
data_pivot_children = data.pivot_table(index='children', columns='debt', 
                                       values='gender', aggfunc='count', fill_value=0)
data_pivot_children

#создаем сводную таблицу, которая покажет кол-во клиентов по кол-ву детей
#и наличию задолженности по кредиту

debt          0     1
children             
0         13028  1063
1          4410   445
2          1858   194
3           303    27
4            37     4
5             9     0

Добавим в сводную таблицу новый столбец, который покажет отношение числа клиентов с задолженностью по кредиту к общему числу клиентов в зависимости от количества детей:

In [35]:
data_pivot_children['ratio'] = (data_pivot_children[1] /
                              (data_pivot_children[0] + data_pivot_children[1]) * 100)
data_pivot_children

#умножаем результат деления на 100, чтобы получить соотношение в процентах

debt          0     1     ratio
children                       
0         13028  1063  7.543822
1          4410   445  9.165808
2          1858   194  9.454191
3           303    27  8.181818
4            37     4  9.756098
5             9     0  0.000000

**Вывод**

Судя по результатам, некоторая зависимость между количеством детей и наличием задолженности по кредиту все же есть. Процент бездетных клиентов с задолженностью ниже аналогичного показателя у клиентов с детьми, учитывая еще и тот факт, что в целом таких клиентов гораздо больше, чем остальных. 
Результаты по клиентам с тремя детьми и больше малопоказательны из-за небольшой выборки. Однако если объединить многодетных клиентов в отдельную категорию, процент наличия задолженности по кредиту у них все равно будет больше, чем у клиентов без детей.   

In [36]:
result = ((data_pivot_children.loc[3,1] + data_pivot_children.loc[4,1]) /
         (data_pivot_children.loc[3,0] + data_pivot_children.loc[3,1] + 
          data_pivot_children.loc[4,0] + data_pivot_children.loc[4,1] + 
          data_pivot_children.loc[5,0]) * 100)
result

### понимаю, что значения можно было ввести вручную, но решила отработать индексацию

8.157894736842106

### Зависимость между семейным положением и возвратом кредита в срок

Определим, есть ли зависимость между семейным положением и возвратом кредита в срок с помощью сводной таблицы. 

In [37]:
data_pivot_family = data.pivot_table(index='family_status_id', columns='debt', 
                                     values='gender', aggfunc='count', fill_value=0)
data_pivot_family

#создаем сводную таблицу, которая покажет кол-во клиентов по наличию задолженности 
#по кредиту в зависимости от семейного статуса

debt                  0    1
family_status_id            
0                 11362  928
1                  3754  385
2                   892   63
3                  1109   84
4                  2528  273

Для того, чтобы понять, какой id какое семейное положение обозначает, объединим полученную сводную таблицу с ранее созданной таблицей-"словарем" с помощью метода `merge()`:

In [38]:
data_subcategory = data_pivot_family.merge(data_family, on='family_status_id', how='left')
data_subcategory

#присоединяем таблицу data_family к таблице data_pivot_family 
#на основании общего столбца 'family_status_id'

family_status_id      0    1          family_status
0                 0  11362  928        женат / замужем
1                 1   3754  385       гражданский брак
2                 2    892   63         вдовец / вдова
3                 3   1109   84              в разводе
4                 4   2528  273  Не женат / не замужем

Теперь рассчитаем соотношение клиентов с задолженностью и общего числа клиентов для каждой категории семейного положения и добавим результаты в отдельный столбец.

In [39]:
data_subcategory['ratio'] = data_subcategory[1] / (data_subcategory[0] + data_subcategory[1]) * 100
data_subcategory.sort_values(by='ratio', ascending=False)

#умножаем результат деления на 100, чтобы получить соотношение в процентах
#выводим таблицу на экран, отсортировав по уменьшению значений столбца 'ratio'

family_status_id      0    1          family_status     ratio
4                 4   2528  273  Не женат / не замужем  9.746519
1                 1   3754  385       гражданский брак  9.301764
0                 0  11362  928        женат / замужем  7.550854
3                 3   1109   84              в разводе  7.041073
2                 2    892   63         вдовец / вдова  6.596859

**Вывод**

Зависимость между семейным положением и возвратом кредита в срок определенно прослеживается. Больше всего доля задолженности по кредиту среди клиентов, не связанных узами брака и состоящих в гражданском браке. Соответственно те, кто состоит в официальном браке или состоял ранее, реже являются должниками.

### Зависимость между уровнем дохода и возвратом кредита в срок

С помощью сводной таблицы узнаем существует ли зависимость между уровнем дохода и возвратом кредита в срок:

In [40]:
data_pivot_income = data.pivot_table(index='total_income_category', columns='debt', 
                                     values='gender', aggfunc='count', fill_value=0)
data_pivot_income

#создаем сводную таблицу, которая покажет кол-во клиентов по наличию задолженности 
#по кредиту в зависимости от уровня дохода


debt                       0     1
total_income_category             
A                         23     2
B                       4667   354
C                      14606  1354
D                        329    21
E                         20     2

Добавим в сводную таблицу новый столбец, который покажет отношение числа клиентов с задолженностью по кредиту к общему числу клиентов по каждой категории дохода:

In [41]:
data_pivot_income['ratio'] = data_pivot_income[1] / (data_pivot_income[0] + data_pivot_income[1]) * 100
data_pivot_income.sort_values(by='ratio', ascending=False)

#умножаем результат деления на 100, чтобы получить соотношение в процентах
#выводим таблицу на экран, отсортировав по уменьшению значений столбца 'ratio'

debt                       0     1     ratio
total_income_category                       
E                         20     2  9.090909
C                      14606  1354  8.483709
A                         23     2  8.000000
B                       4667   354  7.050388
D                        329    21  6.000000

**Вывод**

Некоторая зависимость между уровнем дохода и возвратом кредита в срок существует. Судя по результатам в категориях B и C, чем выше доход клиента(B), тем меньше доля задолженностей (почти на 1,5% по сравнению с категорией C). Однако, чтобы с полной уверенностью говорить о корреляции между уровнем дохода и наличием задолженности по кредиту, необходимо расширить выборку и дополнить данные по клиентам с категорией дохода A, D и E.

### Зависимость между целью получения кредита и его возвратом в срок

Чтобы ответить на вопрос о наличии зависимости снова обратимся к сводной таблице:

In [42]:
data_pivot_purpose = data.pivot_table(index='purpose_category', columns='debt', 
                                      values='gender', aggfunc='count', fill_value=0)
data_pivot_purpose

#создаем сводную таблицу, которая покажет кол-во клиентов по наличию задолженности 
#по кредиту в зависимости от цели получения кредита

debt                         0    1
purpose_category                   
операции с автомобилем    3889  401
операции с недвижимостью  9995  780
получение образования     3629  369
проведение свадьбы        2132  183

Рассчитаем соотношение клиентов с задолженностью и общего числа клиентов в зависимости от цели получения кредита и запишем результаты в новый столбец.

In [43]:
data_pivot_purpose['ratio'] = (data_pivot_purpose[1] / (data_pivot_purpose[0] + 
                                                       data_pivot_purpose[1]) * 100)
data_pivot_purpose.sort_values(by='ratio', ascending=False)

#умножаем результат деления на 100, чтобы получить соотношение в процентах
#выводим таблицу на экран, отсортировав по уменьшению значений столбца 'ratio'

debt                         0    1     ratio
purpose_category                             
операции с автомобилем    3889  401  9.347319
получение образования     3629  369  9.229615
проведение свадьбы        2132  183  7.904968
операции с недвижимостью  9995  780  7.238979

**Вывод**

Гипотеза о зависимости между целью получения кредита и его возвратом в срок подтверждается. Чаще всего задолженность возникает при получении кредита на различные операции с автомобилем и получение образования. Кредиты, на операции с недвижимостью, напротив, чаще всего погашаются вовремя. 

## Итоги исследования

Мы проверили четыре гипотезы и установили:
1. Зависимость между количеством детей и возвратом кредита в срок.

Гипотеза подтвердилась, зависимость существует. Чем больше у клиента банка детей, тем больше вероятность просрочки по кредиту. 

2. Зависимость между семейным положением и возвратом кредита в срок.

Гипотеза также подтвердилась. Как правило, у клиентов, не состоящих в браке или находящихся в гражданском браке, чаще есть задолженность по кредиту. 

3. Зависимость между уровнем дохода и возвратом кредита в срок.

Некоторая зависимость между уровнем дохода и возвратом кредита в срок существует. Клиенты с более низким доходом показывают больший процент задолженности по кредиту. Однако, прежде чем учитывать этот результат при построении модели кредитного скоринга, есть смысл расширить выборку и дополнить данные по клиентам с другими категориями дохода.

4. Зависимость между целью получения кредита и его возвратом в срок.

Четвертая гипотеза подтвердилась. Чаще всего возвращают в срок кредиты, полученные на операции с недвижимостью. Хуже всего дело обстоит с кредитами, взятыми на различные операции с автомобилем и получение образования. 